src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs

```diff
@@ -107,84 +136,97 @@ public async Task BlazorServerTemplateWorks_IndividualAuth(bool useLocalDB)
+                    await TestBasicNavigation(page);
+                    await page.CloseAsync();
+                }
+                else
                 {
-                    aspNetProcess.VisitInBrowser(Browser);
-                    TestBasicNavigation();
+                    Assert.False(
+                        TestHelpers.TryValidateBrowserRequired(
+                            browserKind,
+                            isRequired: !Fixture.BrowserManager.IsExplicitlyDisabled(browserKind),
+                            out var errorMessage),
+                        errorMessage);
                 }
             }
         }
 
-        private void TestBasicNavigation()
+        private async Task TestBasicNavigation(IPage page)
         {
-            var retries = 3;
-            var connected = false;
-            do
-            {
-                try
-                {
-                    Browser.Contains("Information: WebSocket connected to",
-                        () => string.Join(Environment.NewLine, Browser.GetBrowserLogs(LogLevel.Info).Select(b => b.Message)));
-                    connected = true;
-                }
-                catch (TimeoutException) when(retries-- > 0)
-                {
-                    Browser.Navigate().Refresh();
-                }
-            } while (!connected && retries > 0);
-
-
-            Browser.Exists(By.TagName("ul"));
+            await page.QuerySelectorAsync("ul");
             // <title> element gets project ID injected into it during template execution
-            Browser.Equal(Project.ProjectName.Trim(), () => Browser.Title.Trim());
+            Assert.Equal(Project.ProjectName.Trim(), (await page.GetTitleAsync()).Trim());
 
             // Initially displays the home page
-            Browser.Equal("Hello, world!", () => Browser.FindElement(By.TagName("h1")).Text);
+            Assert.Equal("Hello, world!", await page.GetTextContentAsync("h1"));
 
             // Can navigate to the counter page
-            Browser.Click(By.PartialLinkText("Counter"));
-            Browser.Contains("counter", () => Browser.Url);
-            Browser.Equal("Counter", () => Browser.FindElement(By.TagName("h1")).Text);
+
+            await Task.WhenAll(
+                page.WaitForNavigationAsync("**/counter"),
+                page.WaitForSelectorAsync("text=Current count: 0"),
+                page.ClickAsync("text=Counter"));
+
+            Assert.Equal("Counter", await page.GetTextContentAsync("h1"));
 
             // Clicking the counter button works
-            Browser.Equal("Current count: 0", () => Browser.FindElement(By.CssSelector("h1 + p")).Text);
-            Browser.Click(By.CssSelector("p+button"));
-            Browser.Equal("Current count: 1", () => Browser.FindElement(By.CssSelector("h1 + p")).Text);
+            Assert.Equal("Current count: 0", await page.GetTextContentAsync("h1+p"));
+            await page.WaitForTimeoutAsync(1000);
```

Comment:
What's this for?

Reply:
I didn't clean up this part of the script, I've cleaned it up.